core/src/transport.rs

```diff
@@ -149,10 +149,24 @@ pub trait Transport {
         cx: &mut Context<'_>,
     ) -> Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>>;
 
-    /// Performs a transport-specific mapping of an address `observed` by
-    /// a remote onto a local `listen` address to yield an address for
-    /// the local node that may be reachable for other peers.
-    fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr>;
+    /// Performs a transport-specific mapping of an address `observed` by a remote onto a
+    /// local `listen` address to yield an address for the local node that may be reachable
+    /// for other peers.
+    ///
+    /// This is relevant for transports where Network Address Translation (NAT) can occur
+    /// so that e.g. the peer is observed at a different IP than the IP of the local
+    /// listening address. See also [`address_translation`][crate::address_translation].
+    ///
+    /// Within [`libp2p_swarm::Swarm`][libp2p_swarm::Swarm] this is used when extending the
+    /// listening addresses of the local peer with external addresses observed by remote peers.
+    /// On transports where this is not relevant (i.e. no NATs are present) `None` should be
+    /// returned for the sake of de-duplication.
+    ///
+    /// Note: if the listen or observed address is not a valid address of this transport,
+    /// `None` should be returned as well.
+    fn address_translation(&self, _listen: &Multiaddr, _observed: &Multiaddr) -> Option<Multiaddr> {
+        None
+    }
```

Comment:
```suggestion
    fn address_translation(&self, _listen: &Multiaddr, _observed: &Multiaddr) -> Option<Multiaddr>;
```

Implementing `address_translation` for combinator `Transport` implementations like `OrTransport` is relevant. I suggest we don't provide a default implementation. In case we write another combinator `Transport` implementation like `OrTransport`, we won't miss implementing it.

Reply:
Agree that explicit is better. If `Transport` were something that is implemented a lot outside of this repo it'd be different but it isn't really so it seems like a footgun to me.

Reply:
Dopped the commit.